eventsources/sources/calendar/start.go

```diff
@@ -88,45 +173,82 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 		}
 		lastT = lastT.In(location)
 	}
+	sendEventFunc := func(tx time.Time) error {
+		eventData := &events.CalendarEventData{
+			EventTime:   tx.String(),
+			UserPayload: calendarEventSource.UserPayload,
+			Metadata:    calendarEventSource.Metadata,
+		}
+		payload, err := json.Marshal(eventData)
+		if err != nil {
+			el.log.Error("failed to marshal the event data", zap.Error(err))
+			// no need to continue as further event payloads will suffer same fate as this one.
+			return errors.Wrapf(err, "failed to marshal the event data for event source %s / %s", el.GetEventSourceName(), el.GetEventName())
+		}
+		el.log.Info("dispatching calendar event...")
+		err = dispatch(payload)
+		if err != nil {
+			el.log.Error("failed to dispatch calendar event", zap.Error(err))
+			return errors.Wrapf(err, "failed to dispatch calendar event")
+		}
+		if el.eventPersistence != nil && el.eventPersistence.IsEnabled() {
+			event := persist.Event{EventKey: el.getPersistenceKey(), EventPayload: string(payload)}
+			err = el.eventPersistence.Save(&event)
+			if err != nil {
+				el.log.Error("failed to persist calendar event", zap.Error(err))
+			}
+		}
+		return nil
+	}
 
+	el.log.Info("Calendar event start time:", zap.Any("Time", lastT.Format(time.RFC822)))
 	for {
 		t := next(lastT)
+
+		// Catchup scenario
+		// Trigger the event immediately if the current schedule time is earlier then
+		if time.Now().After(t) {
+			el.log.Info("triggering catchup events", zap.Any(logging.LabelTime, t.UTC().String()))
+			lastT = t
+			if location != nil {
+				lastT = lastT.In(location)
+			}
+			err = sendEventFunc(t)
+			if err != nil {
+				el.log.Error("failed to send calendar event", zap.Error(err))
+			}
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
 		timer := time.After(time.Until(t))
-		log.Info("expected next calendar event", zap.Any(logging.LabelTime, t.UTC().String()))
+		el.log.Info("expected next calendar event", zap.Any(logging.LabelTime, t.UTC().String()))
 		select {
 		case tx := <-timer:
 			lastT = tx
 			if location != nil {
 				lastT = lastT.In(location)
 			}
-			response := &events.CalendarEventData{
-				EventTime:   tx.String(),
-				UserPayload: calendarEventSource.UserPayload,
-				Metadata:    calendarEventSource.Metadata,
-			}
-			payload, err := json.Marshal(response)
-			if err != nil {
-				log.Error("failed to marshal the event data", zap.Error(err))
-				// no need to continue as further event payloads will suffer same fate as this one.
-				return errors.Wrapf(err, "failed to marshal the event data for event source %s / %s", el.GetEventSourceName(), el.GetEventName())
-			}
-			log.Info("dispatching calendar event...")
-			err = dispatch(payload)
+			err = sendEventFunc(tx)
 			if err != nil {
-				log.Error("failed to dispatch calendar event", zap.Error(err))
+				el.log.Error("failed to send calendar event", zap.Error(err))
```

Comment:
If `err != nil` due to sending message to EventBus failure, because `lastT = tx` is set above, it will not catch up the failed one. This logic works when there's no persistence, but it's a bug when persistence is enabled.

Shouldn't `lastT` be always read from the configmap when persistence is enabled?

Reply:
Good catch

Reply:
How is this issue addressed?

Reply:
Again, how is this addressed?